src/cache.ts

```diff
@@ -93,18 +97,25 @@ export class ListWatch<T extends KubernetesObject> implements ObjectCache<T>, In
     }
 
     public get(name: string, namespace?: string): T | undefined {
-        return this.objects.find(
-            (obj: T): boolean => {
-                return obj.metadata!.name === name && (!namespace || obj.metadata!.namespace === namespace);
-            },
-        );
+        const nsObjects = this.objects.get(namespace || '');
+        if (nsObjects) {
+            return nsObjects.get(name);
+        }
     }
 
     public list(namespace?: string | undefined): ReadonlyArray<T> {
         if (!namespace) {
-            return this.objects;
+            const allObjects: T[] = [];
```

Comment:
If it is unacceptable for performance to do this concatting on each call I can modify it to store a secondary array with all of the objects. However, based on using the go client, I'm not sure that I'd expect this `list` to return everything in the cache without a namespace. Like my comment `get` functionality my expectation is that `list` with no namespace would only return objects that don't have namespaces like pvcs and namespaces rather than all objects. But a number of tests rely on the later behavior and this would be a breaking change.

Reply:
The standard client libraries have a 'listAllResources' function. Given that a catch is tied to a specific Kubernetes type, if I create a watch for a namespaced resource (e.g. a Pod) I would expect that I can get a list of all pods regardless of namespace. This is useful (for example) when writing a scheduler, where I really do want to watch all pods in all namespaces without having to create N different watches for N namespaces (and manage what happens when namespaces come and go)

So I would prefer to retain the pre-existing behavior. From a performance standpoint though, I think it is fine to just concatenate all of the various lists together into a single list.